src/System.Slices/src/System/Contract.cs

```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static void RequiresInRange(int start, int length)
         {
-            if (!(start >= 0 && start < length))
+            // explicit unchecked block is used in order to avoid OverflowException when casting 
+            // negative int to unsigned int when the code is compiled with /checked option
+            unchecked
             {
-                throw new ArgumentOutOfRangeException();
+                // use unsigned int to reduce range check by one ( >= 0 )
+                if ((uint)start >= (uint)length)
```

Comment:
I did some calculations. The minimal value that can be created out of casting signed negative integer to unsigned integer is 2,147,483,648. .NET has limit for single dimension array size which is 2,147,483,591 for bytes array and 2,146,435,071  for array of other types.

So the minimal value that can be generated is always bigger that the .NET's array size limit so the single comparision with Length should be always safe. 

What do you guys think?